dojo/problem/helper.py

```diff
@@ -0,0 +1,137 @@
+import json
+import os
+import requests
+from requests.packages.urllib3.exceptions import InsecureRequestWarning
+
+from dojo.models import Problem, Finding
+
+import logging
+logger = logging.getLogger(__name__)
+
+CONFIG_FILE = os.path.join(os.path.dirname(__file__), 'config.json')
+CACHED_JSON_FILE = os.path.join('/app/media', 'cached_disambiguator.json')
+
+SEVERITY_ORDER = {
+    'Critical': 5,
+    'High': 4,
+    'Medium': 3,
+    'Low': 2,
+    'Info': 1
+}
+
+def load_config():
+    with open(CONFIG_FILE, 'r') as f:
+        return json.load(f)
+
+def validate_json(data):
+    if not isinstance(data, dict):
+        return False
+    for key, value in data.items():
+        if not isinstance(key, str) or not isinstance(value, list):
+            return False
+        if not all(isinstance(item, str) for item in value):
+            return False
+    return True
+
+def download_json(json_url):
+    response = requests.get(json_url, timeout=5, verify=False)
+    response.raise_for_status()
+    return response.json()
+
+def load_cached_json():
+    try:
+        if os.path.exists(CACHED_JSON_FILE):
+            with open(CACHED_JSON_FILE, 'r') as f:
+                data = json.load(f)
+                if validate_json(data):
+                    return data
+    except (ValueError, json.JSONDecodeError):
+        pass 
+    return None
+
+def save_json_to_cache(data):
+    logger.info('Saving disambiguator JSON to cache')
+    with open(CACHED_JSON_FILE, 'w') as f:
+        json.dump(data, f)
+
+def load_json():
+    try:
+        # Disable SSL warnings
+        requests.packages.urllib3.disable_warnings(InsecureRequestWarning)
+
+        cached_data = load_cached_json()
+        if cached_data:
+            return cached_data
+
+        # Cache is missing or invalid, download and validate
+        config = load_config()
+        json_url = config.get('json_url')
+
+        if json_url:
+            data = download_json(json_url)
+            if validate_json(data):
+                save_json_to_cache(data)
+                return data
+
+        return {}
+
+    except (requests.RequestException, ValueError, json.JSONDecodeError) as e:
+        logger.error('Error loading disambiguator JSON: %s', e)
+        return {}
+
+def extract_script_id(full_id):
+    parts = full_id.split('____')
+    return parts[0] if len(parts) == 2 else None
+
+def find_or_create_problem(finding):
+    data = load_json()
+    script_id = finding.vuln_id_from_tool
+
+    valid_ids_mapping = {
+        key: [extract_script_id(full_id) for full_id in script_ids if extract_script_id(full_id)]
+        for key, script_ids in data.items()
+    }
+
+    for key, valid_ids in valid_ids_mapping.items():
```

Comment:
This performs a O(N) search; the mappings can be stored in a dict mapping script-id to problem-id, which would allow for O(1) access.